src/reactive-membrane.ts

```diff
@@ -140,34 +143,35 @@ export class ReactiveMembrane {
         return unwrap(p);
     }
 
-    private getReactiveState(value: any): ReactiveState {
+    private getReactiveState(value: any, distortedValue: any): ReactiveState {
         const {
             objectGraph,
         } = this;
-        value = unwrap(value);
-        let reactiveState = objectGraph.get(value);
+        let reactiveState = objectGraph.get(distortedValue);
         if (reactiveState) {
             return reactiveState;
         }
         const membrane = this;
         reactiveState = {
             get reactive() {
-                const reactiveHandler = new ReactiveProxyHandler(membrane, value);
+                const reactiveHandler = new ReactiveProxyHandler(membrane, distortedValue);
                 // caching the reactive proxy after the first time it is accessed
-                const proxy = new Proxy(createShadowTarget(value), reactiveHandler);
+                const proxy = new Proxy(createShadowTarget(distortedValue), reactiveHandler);
+                registerProxy(proxy, value);
                 ObjectDefineProperty(this, 'reactive', { value: proxy });
                 return proxy;
             },
             get readOnly() {
-                const readOnlyHandler = new ReadOnlyHandler(membrane, value);
+                const readOnlyHandler = new ReadOnlyHandler(membrane, distortedValue);
                 // caching the readOnly proxy after the first time it is accessed
-                const proxy = new Proxy(createShadowTarget(value), readOnlyHandler);
+                const proxy = new Proxy(createShadowTarget(distortedValue), readOnlyHandler);
+                registerProxy(proxy, value);
                 ObjectDefineProperty(this, 'readOnly', { value: proxy });
                 return proxy;
             }
         } as ReactiveState;
 
-        objectGraph.set(value, reactiveState);
+        objectGraph.set(distortedValue, reactiveState);
```

Comment:
Why do we want to use `distorderValue` here instead of `value`?

Reply:
Because of two reasons
1. The handler's behavior is based on the distorted value. So the object graph should have a 1:1 relationship with the value that it is encapsulating.
2. Handles both deterministic and non-deterministic distortion
   a. If distortion is deterministic, then there is a 1:1 mapping between value <-> distortedValue. So we can use either, it makes no difference.
  b. If distortion is non-deterministic, 
       And, the object graph is keyed by `value`, then we will always produce the same ReactiveState for a given value. We are removing the non-deterministic nature of the distortion. Which may be or may not be desirable.